azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -969,223 +930,336 @@ def do_retry():
+    DISCONNECTING = "DISCONNECTING"  # Client is in the process of disconnecting
+    REAUTHORIZING = "REAUTHORIZING"  # Client is in the process of reauthorizing (disconn->conn)
 
 
 class ReconnectStage(PipelineStage):
+
+    progress_states = [
+        ReconnectState.CONNECTING,
+        ReconnectState.DISCONNECTING,
+        ReconnectState.REAUTHORIZING,
+    ]
+    connection_ops = [
+        pipeline_ops_base.ConnectOperation,
+        pipeline_ops_base.DisconnectOperation,
+        pipeline_ops_base.ReauthorizeConnectionOperation,
+    ]
+
     def __init__(self):
         super(ReconnectStage, self).__init__()
         self.reconnect_timer = None
-        self.state = ReconnectState.LOGICALLY_DISCONNECTED
-        # never_connected is important because some errors are handled differently the first time
-        # that we're connecting versus later connections.
-        #
-        # For example, if we get a "host not found" the first time we connect, it might mean:
-        # 1. The connection string is wrong, in which case we don't want to retry
-        # or it might mean:
-        # 2. The connection string is correct and the host is just not available (network glitch?)
-        #
-        # If never_connected=True, we don't know if it's a permanent error (#1) or if it's
-        # temporary (#2), so we take the conservative approach and assume it's a permanent error.
-        #
-        # If never_connected=False, we know that we've connected before, so the credentials were
-        # valid at some point in the recent past.  We still don't know with certainty if it's
-        # a temporary error or if it's permanent (the hub may have been deallocated), but it has
-        # worked in the past, so we assume it's a temporary error.
-        self.never_connected = True
-        self.waiting_connect_ops = []
+        self.state = ReconnectState.DISCONNECTED
+        self.state_lock = threading.Lock()
+        self.waiting_ops = queue.Queue()
 
     @pipeline_thread.runs_on_pipeline_thread
     def _run_op(self, op):
-        if isinstance(op, pipeline_ops_base.ConnectOperation):
-            if self.state == ReconnectState.WAITING_TO_RECONNECT:
+        # Connection Retry Enabled
+        if self.pipeline_root.pipeline_configuration.connection_retry:
+
+            # NOTE: manual lock is used here rather than with context manager for structural
+            # reasons. We want the lock while checking the conditional, but not for what occurs
+            # as a result of the condition.
+            self.state_lock.acquire()
+
+            # If receiving a connection op while one is already in progress, wait for the current
+            # one to finish. This is kind of like a ConnectionLockStage, but inside this one.
+            # It has to happen here because just relying on a ConnectionLockStage before or after
+            # in the pipeline is insufficient, given that operations can spawn in this stage.
```

Comment:
Does this mean we can get rid of ConnectionLockStage?  If not, why not?

Reply:
I think the ConnectionLockStage has value, but may need some adjustments. However, I didn't want to expand the scope of this PR any further. A little redundancy for now isn't the worst thing.